File: templates/playwright/.grader/summarize.js
// Reads grader-results-raw.json (Playwright JSON reporter format)
// Writes grader-results.json (boilerplate standard format)
// Writes grader-report.md (PR comment)

const fs   = require('fs');
const path = require('path');

const ROOT     = process.cwd();
const RAW_FILE = path.join(ROOT, 'grader-results-raw.json');
const OUT_JSON = path.join(ROOT, 'grader-results.json');
const OUT_MD   = path.join(ROOT, 'grader-report.md');

// ---------------------------------------------------------------------------
// Rubric — points per category.
// These values MUST match the per-file point totals configured in GitHub
// Classroom (one autograder test per file, e.g. structure.spec.ts → 6 pts).
// Add or adjust categories to match the specific lab's test files.
// ---------------------------------------------------------------------------

const CATEGORY_POINTS = {
    'Structure': 6,
    'Execution': 17,
    'Quality':   6,
};

const TOTAL_POINTS = Object.values(CATEGORY_POINTS).reduce((a, b) => a + b, 0);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function extractHint(spec) {
    const msg = spec.tests?.[0]?.results?.[0]?.error?.message ?? '';
    if (!msg) return null;

    // Strip ANSI color codes
    const clean = msg.replace(/\x1b\[[0-9;]*m/g, '');
    const lines = clean.split('\n').map(l => l.trim()).filter(l => l.length > 0);

    // Playwright puts the custom expect message on the first line — prefer it
    if (lines[0] && !lines[0].startsWith('expect(') && !lines[0].startsWith('Error:')) {
        return lines[0].slice(0, 400);
    }

    // Fall back: first 3 lines joined
    return lines.slice(0, 3).join(' · ').slice(0, 400) || null;
}

function progressBar(score, total) {
    if (total === 0) return '`░░░░░░░░░░`';
    const filled = Math.round((score / total) * 10);
    return '`' + '█'.repeat(filled) + '░'.repeat(10 - filled) + '`';
}

function statusMessage(pct) {
    if (pct === 100) return 'Perfect score! Excellent work!';
    if (pct >= 80)   return 'Almost there — just a few more to fix!';
    if (pct >= 50)   return 'Good progress, keep going!';
    if (pct >= 20)   return "You've started! Keep building.";
    return 'Time to write some code!';
}

function catIcon(cat) {
    if (cat.score === cat.total) return '✅';
    if (cat.score === 0)         return '❌';
    return '⚠️';
}

function detectStudentCode() {
    const srcDir = path.join(ROOT, 'src');
    try {
        return hasFiles(srcDir, f => !f.endsWith('.gitkeep'));
    } catch {
        return false;
    }
}

function hasFiles(dir, predicate) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        if (entry.isDirectory()) {
            if (hasFiles(path.join(dir, entry.name), predicate)) return true;
        } else if (predicate(entry.name)) {
            return true;
        }
    }
    return false;
}

function renderAiSection(ai) {
    if (!ai.evaluated) {
        const reason = ai.reason || 'AI Review: not configured.';
        return `\n---\n\n> ${reason}\n\n`;
    }
    const artifactLabel = path.basename(ai.artifact || 'artifact');
    const total = ai.total_score ?? 0;
    const maxS = ai.max_score ?? 0;
    let md = `\n---\n\n### AI Review — ${artifactLabel}   ${total} / ${maxS}\n\n`;
    md += '> *Coaching feedback — not included in your automated score.*\n\n';
    const summary = ai.summary || '';
    if (summary) {
        md += `> **Overall:** ${summary}\n\n`;
    }
    const criteria = ai.criteria || [];
    criteria.forEach((criterion, i) => {
        const openAttr = i === 0 ? ' open' : '';
        md += `<details${openAttr}>\n`;
        md += `<summary><b>${criterion.name}</b> — ${criterion.score} / ${criterion.max}</summary>\n\n`;
        md += `**Strength:** ${criterion.strength || ''}\n\n`;
        md += `**Gap:** ${criterion.gap || ''}\n\n`;
        md += `**Action:** ${criterion.action || ''}\n\n`;
        md += '</details>\n\n';
    });
    return md;
}

// ---------------------------------------------------------------------------
// Read raw results
// ---------------------------------------------------------------------------

let raw;
try {
    raw = JSON.parse(fs.readFileSync(RAW_FILE, 'utf-8'));
} catch (e) {
    console.error('Could not read grader-results-raw.json:', e.message);
    process.exit(1);
}

// ---------------------------------------------------------------------------
// Parse Playwright JSON format: suites[file].suites[describe].specs[]
// ---------------------------------------------------------------------------

const categories = [];

for (const fileSuite of raw.suites ?? []) {
    for (const describeSuite of fileSuite.suites ?? []) {
        const checks = (describeSuite.specs ?? []).map(spec => ({
            name:   spec.title,
            passed: spec.ok,
            hint:   spec.ok ? null : extractHint(spec),
        }));

        const score    = checks.filter(c => c.passed).length;
        const total    = checks.length;
        const maxPts   = CATEGORY_POINTS[describeSuite.title] ?? 0;
        const points   = total > 0 && maxPts > 0
            ? Math.round((score / total) * maxPts * 10) / 10
            : 0;

        categories.push({
            name:      describeSuite.title,
            score,
            total,
            points,
            maxPoints: maxPts,
            checks,
        });
    }
}

const totalPtsEarned = Math.round(categories.reduce((sum, c) => sum + c.points, 0) * 10) / 10;
const durationMs     = raw.stats?.duration ?? 0;
const labName        = path.basename(ROOT);

// ---------------------------------------------------------------------------
// Write grader-results.json
// ---------------------------------------------------------------------------

const results = {
    lab:        labName,
    score:      totalPtsEarned,
    total:      TOTAL_POINTS,
    duration:   durationMs,
    categories,
};

fs.writeFileSync(OUT_JSON, JSON.stringify(results, null, 2));
console.log(`grader-results.json written (${totalPtsEarned}/${TOTAL_POINTS} pts)`);

// ---------------------------------------------------------------------------
// Write grader-report.md
// ---------------------------------------------------------------------------

const pct            = TOTAL_POINTS > 0 ? Math.round((totalPtsEarned / TOTAL_POINTS) * 100) : 0;
const hasStudentCode = detectStudentCode();

let md = '## Lab Grader Results\n\n';
md += `> **Score: ${totalPtsEarned} / ${TOTAL_POINTS} pts (${pct}%)** — ${statusMessage(pct)}\n`;
md += `> ${progressBar(totalPtsEarned, TOTAL_POINTS)}\n\n`;
md += '---\n\n';

// Summary table
md += '### Score Breakdown\n\n';
md += '| | Category | Points | Checks |\n';
md += '|---|---|---|---|\n';
for (const cat of categories) {
    const icon      = catIcon(cat);
    const checksStr = cat.score < cat.total ? `${cat.score}/${cat.total}` : 'all passing';
    md += `| ${icon} | **${cat.name}** | ${cat.points} / ${cat.maxPoints} | ${checksStr} |\n`;
}
md += '\n---\n\n';

if (!hasStudentCode) {
    md += '### Getting Started\n\n';
    md += "*It looks like you haven't added any code yet — here's where to begin:*\n\n";
    md += '1. Read `docs/instructions.md` for the full exercise description\n';
    md += '2. Add your code inside the `src/` directory\n';
    md += '3. Run `npm test` locally to verify before pushing\n\n';
    md += '---\n\n';
}

// Per-category detail sections
md += '### Details\n\n';

for (const cat of categories) {
    const icon      = catIcon(cat);
    const ptsLabel  = `${cat.points} / ${cat.maxPoints} pts`;
    const chkLabel  = `${cat.score}/${cat.total} checks`;
    const failing   = cat.checks.filter(c => !c.passed);

    md += '<details>\n';
    md += `<summary>${icon} <b>${cat.name}</b> — ${ptsLabel} — ${chkLabel}</summary>\n\n`;

    md += '| Check | Status | Hint |\n';
    md += '|---|---|---|\n';
    for (const check of cat.checks) {
        const status = check.passed ? '✅ Pass' : '❌ Fail';
        const hint   = check.hint ?? '';
        md += `| ${check.name} | ${status} | ${hint} |\n`;
    }

    if (failing.length > 0) {
        md += '\n**What to fix:**\n';
        for (const check of failing) {
            const hintText = check.hint ? ` — ${check.hint}` : '';
            md += `- \`${check.name}\`${hintText}\n`;
        }
    }

    md += '\n</details>\n\n';
}

md += '---\n\n';
md += '### Resources\n\n';
md += '- [Playwright Locators](https://playwright.dev/docs/locators)\n';
md += '- [Page Object Model](https://playwright.dev/docs/pom)\n';
md += '- [Assertions reference](https://playwright.dev/docs/test-assertions)\n';
md += '- Lab instructions: `docs/instructions.md`\n';

const aiFile = path.join(ROOT, 'ai-feedback.json');
let ai = { evaluated: false, reason: 'AI Review: not configured — set PORTKEY_API_KEY to enable.' };
if (fs.existsSync(aiFile)) {
    try { ai = JSON.parse(fs.readFileSync(aiFile, 'utf-8')); } catch (_) {}
}
md += renderAiSection(ai);

fs.writeFileSync(OUT_MD, md);
console.log('grader-report.md written');

File: templates/python-copilot/.grader/summarize.py
"""
Reads grader-results-raw.json (pytest-json-report format)
Reads ai-feedback.json (from ai_evaluator.py)
Writes grader-results.json  (boilerplate standard format)
Writes grader-report.md     (PR comment)
"""

import json
import os
import sys
from pathlib import Path


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def progress_bar(passed: int, total: int) -> str:
    if total == 0:
        return "`░░░░░░░░░░`"
    filled = round((passed / total) * 10)
    return "`" + "█" * filled + "░" * (10 - filled) + "`"


def status_message(p: int) -> str:
    if p == 100: return "Perfect score! Excellent work!"
    if p >= 80:  return "Almost there — just a few more to fix!"
    if p >= 50:  return "Good progress, keep going!"
    if p >= 20:  return "You've started! Keep building."
    return "Time to write some code!"


def has_student_code(root: str) -> bool:
    src = os.path.join(root, "src")
    if not os.path.isdir(src):
        return False
    for r, _, files in os.walk(src):
        for f in files:
            if not f.endswith(".gitkeep"):
                return True
    return False


def render_ai_section(ai: dict) -> str:
    """Render the AI Review section of the PR comment."""
    if not ai.get("evaluated"):
        reason = ai.get("reason", "AI Review: not configured.")
        return f"\n---\n\n> {reason}\n\n"

    artifact = ai.get("artifact", "")
    artifact_label = Path(artifact).name if artifact else "artifact"
    total = ai.get("total_score", 0)
    max_s = ai.get("max_score", 0)

    md = f"\n---\n\n### AI Review — {artifact_label}   {total} / {max_s}\n\n"
    md += "> *Coaching feedback — not included in your automated score.*\n\n"
    summary = ai.get("summary", "")
    if summary:
        md += f"> **Overall:** {summary}\n\n"
    for i, criterion in enumerate(ai.get("criteria", [])):
        open_attr = " open" if i == 0 else ""
        md += f"<details{open_attr}>\n"
        md += f"<summary><b>{criterion['name']}</b> — {criterion['score']} / {criterion['max']}</summary>\n\n"
        md += f"**Strength:** {criterion.get('strength', '')}\n\n"
        md += f"**Gap:** {criterion.get('gap', '')}\n\n"
        md += f"**Action:** {criterion.get('action', '')}\n\n"
        md += "</details>\n\n"
    return md


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    root = str(Path.cwd())
    raw_file = os.path.join(root, "grader-results-raw.json")
    ai_file  = os.path.join(root, "ai-feedback.json")
    out_json = os.path.join(root, "grader-results.json")
    out_md   = os.path.join(root, "grader-report.md")

    try:
        with open(raw_file, "r") as f:
            raw = json.load(f)
    except Exception as e:
        print(f"Could not read grader-results-raw.json: {e}")
        sys.exit(1)

    # -----------------------------------------------------------------------
    # Parse pytest-json-report format
    # -----------------------------------------------------------------------

    categories: dict[str, dict] = {}

    for test in raw.get("tests", []):
        node   = test["nodeid"]
        parts  = node.split("::")
        group  = parts[1] if len(parts) >= 3 else parts[0]
        name   = parts[-1].replace("_", " ").capitalize()
        passed = test["outcome"] == "passed"

        hint = None
        if not passed:
            msg = (test.get("call") or {}).get("longrepr", "")
            if msg:
                hint = " · ".join(
                    line.strip() for line in str(msg).splitlines()
                    if line.strip()
                )[:200]

        if group not in categories:
            categories[group] = {"name": group, "score": 0, "total": 0, "checks": []}

        categories[group]["checks"].append({"name": name, "passed": passed, "hint": hint})
        categories[group]["total"] += 1
        if passed:
            categories[group]["score"] += 1

    cats         = list(categories.values())
    total_score  = sum(c["score"] for c in cats)
    total_checks = sum(c["total"] for c in cats)
    lab_name     = os.path.basename(root)

    # -----------------------------------------------------------------------
    # Write grader-results.json
    # -----------------------------------------------------------------------

    results = {
        "lab":        lab_name,
        "score":      total_score,
        "total":      total_checks,
        "duration":   raw.get("duration", 0),
        "categories": cats,
    }

    with open(out_json, "w") as f:
        json.dump(results, f, indent=2)
    print(f"grader-results.json written ({total_score}/{total_checks})")

    # -----------------------------------------------------------------------
    # Write grader-report.md
    # -----------------------------------------------------------------------

    pct = round((total_score / total_checks) * 100) if total_checks > 0 else 0

    md  = "## Lab Grader Results\n\n"
    md += f"> **Score: {total_score} / {total_checks} checks passed ({pct}%)** — {status_message(pct)}\n"
    md += f"> {progress_bar(total_score, total_checks)}\n\n"
    md += "---\n\n"

    if not has_student_code(root):
        md += "### Getting Started\n\n"
        md += "*It looks like you haven't added any code yet — here's where to begin:*\n\n"
        md += "1. Read `docs/instructions.md` for the full exercise description\n"
        md += "2. Add your code inside the `src/` directory\n"
        md += "3. Run `pytest` locally to verify before pushing\n\n"
        md += "---\n\n"

    for cat in cats:
        icon = "✅" if cat["score"] == cat["total"] else ("❌" if cat["score"] == 0 else "⚠️")
        md += "<details>\n"
        md += f"<summary>{icon} {cat['name']} — {cat['score']}/{cat['total']} passed</summary>\n\n"
        md += "| Check | | Hint |\n"
        md += "|---|---|---|\n"
        for check in cat["checks"]:
            icon2 = "✅" if check["passed"] else "❌"
            md += f"| {check['name']} | {icon2} | {check['hint'] or ''} |\n"
        md += "\n</details>\n\n"

    md += "---\n\n"
    md += "### Resources\n\n"
    md += "- [GitHub Copilot docs](https://docs.github.com/en/copilot)\n"
    md += "- [Python pytest docs](https://docs.pytest.org/)\n"
    md += "- Lab instructions: `docs/instructions.md`\n"

    ai_data = (
        json.loads(open(ai_file).read())
        if os.path.exists(ai_file)
        else {"evaluated": False, "reason": "AI Review: not configured — set PORTKEY_API_KEY to enable."}
    )
    md += render_ai_section(ai_data)

    with open(out_md, "w") as f:
        f.write(md)
    print("grader-report.md written")


if __name__ == "__main__":
    main()
